contrib/sort: enable f16/f64 sorting on no native FP

seiko2plus opened the pull request:
Add an integer-domain total-order comparator and use it for float16_t/double on targets lacking native fp operations. Preserves IEEE intent (e.g., -0.0 < +0.0); Wires into sort/select traits, removes FP-only guards, and enables basic tests.

This is part of enabling Highway vqsort for NumPy on x86, as intel-x86-sort provides emulation for half-precision

jan-wassenberg commented on hwy/contrib/sort/order-emulate-inl.h:
```diff
@@ -0,0 +1,224 @@
+    // Transform to monotonic space: flip sign for positives, invert for negatives
+    const M is_neg = TestBit(u, sign_bit);
+    const V key = Xor(u, IfThenElse(is_neg, all1, sign_bit));
+    // Boundary detection: +0/-0 swap needs a step of 2 instead of 1
```
Do we really need -0 != 0? Wouldn't native float comparison consider them equal/equivalent?

jan-wassenberg commented on hwy/contrib/sort/sort_unit_test.cc:
```diff
@@ -54,81 +54,114 @@ using detail::Traits128;
+} 
+
+template <class TF, class V, class D = DFromV<V>, HWY_IF_UNSIGNED_D(D)>
+HWY_API Mask<D> IsInfBin(V v) {
```
Can we improve the name somehow? Maybe IsInfWrapper or IsInfEmulated?

jan-wassenberg commented on hwy/contrib/sort/sort_unit_test.cc:
```diff
@@ -54,81 +54,114 @@ using detail::Traits128;
     auto in = hwy::AllocateAligned<T>(num);
     HWY_ASSERT(in);
-    Fill(d, GetLane(Inf(d)), num, in.get());
+    Fill(d, BitCastScalar<T>(ExponentMask<TF>()), num, in.get());
```
Should we also make a wrapper function for Inf() in the same way as IsInf?

jan-wassenberg reviewed (changes requested):
Nice, it's great to make this unconditionally available.
Generally LGTM but the -0 != 0 part is surprising, do we really need that?

jan-wassenberg commented on hwy/contrib/sort/order-emulate-inl.h:
```diff
@@ -0,0 +1,224 @@
+  using TF = typename Base::KeyType;
+
+  HWY_INLINE bool Compare1(const T* a, const T* b) const {
+    return reinterpret_cast<_Asc>(this)->Compare1(b, a);
```
reinterpret_cast and inheritance seems questionable here.
We could instead use a typedef like you have (but no _ prefix because that is reserved by C++), and instead write return AscBase().Compare(d, b, a) - WDYT?

seiko2plus replied:
Make sense, done

jan-wassenberg reviewed (changes requested):
Nice simplifications, I especially like IsEmulatedMinMax, thanks!
Wrapper LGTM.
One remaining concern about inheritance of Descending from Ascending, apart from that good to go.

jan-wassenberg commented:
> I'm going to avoid using operator + in here instead for now to bypass ci errors:
> 
> https://github.com/google/highway/blob/746c513ae5486c9d955f3bbaae7b746b4fb30b23/hwy/contrib/sort/sort_unit_test.cc#L80-L85

Agreed, what we usually do for bf16 and fp16 is to `ConvertScalarTo<float>`, add, then convert back to the actual type.

seiko2plus commented:
To clarify:

- With flush-to-zero enabled, this emulation still handles subnormals, so it’s fine to treat them as integers by the following branches:
https://github.com/google/highway/blob/0bb1960251e6529bab84daa12c6717f46131e397/hwy/contrib/sort/vqsort-inl.h#L1283-L1290

- https://github.com/google/highway/blob/0bb1960251e6529bab84daa12c6717f46131e397/hwy/contrib/sort/vqsort-inl.h#L1434-L1458

- NumPy CI (numpy/numpy#29829) triggered errors, mostly related to select and MMX prefetch. I’ll follow up with fixes after addressing these issues.

seiko2plus commented:
> NumPy CI (https://github.com/numpy/numpy/pull/29829) triggered errors, mostly related to select and MMX prefetch. I’ll follow up with fixes after addressing these issues.

Regarding `VQselect`, the runtime test failure was in NumPy’s test itself—since the order within partitions is undefined.

Now, all tests passed except for a `_mm_prefetch` build error on clang-cl:
```Bash
umpy\_core\src\highway\hwy/cache_control.h(102,3): error: '_mm_prefetch' needs target feature mmx
  102 |   _mm_prefetch(reinterpret_cast<const char*>(p), _MM_HINT_T0);
      |   ^
..\numpy\_core\src\highway\hwy/cache_control.h(102,3): error: '_mm_prefetch' needs target feature mmx
..\numpy\_core\src\highway\hwy/cache_control.h(102,3): error: '_mm_prefetch' needs target feature mmx
..\numpy\_core\src\highway\hwy/cache_control.h(102,3): error: '_mm_prefetch' needs target feature mmx
```

We pass `-mno-mmx` when targeting `AVX512*` due to a previously discovered bug that corrupts the x87 FP stack; I haven’t revisited this since. I don’t know why `clang-cl` requires `MMX`—wouldn’t `SSE` be enough? (Possibly something related to Windows) In any case, I think this PR is ready to go. I’ll follow up with another PR to fix this clang-cl build error by falling back to `__builtin_prefetch` when clang-cl is built with `-mno-mmx`.

jan-wassenberg reviewed (approved):
👍

jan-wassenberg commented:
Internal tests/builds are unfortunately failing due to a build timeout in optimized asan builds for the f16 file. What do you think of splitting it up into separate files for Sort vs Select vs PartialSort? Disabling f16 for asan is probably not desirable.
Any other way we can reduce compile time?

jan-wassenberg commented:
Ah, possible that swapping might slow things down, yes. Unfortunately we don't have control over the build machines.

Good point about PCH. We can least clang header modules in the BUILD file. I'll try that.
